lib/Transforms/IPO/LoopExtractor.cpp

```diff
@@ -31,21 +37,85 @@
 #include "llvm/Transforms/Utils/CodeExtractor.h"
 #include <fstream>
 #include <set>
+
 using namespace llvm;
 
-#define DEBUG_TYPE "loop-extract"
+#define DEBUG_TYPE "sea-loop-extract"
 
 STATISTIC(NumExtracted, "Number of loops extracted");
 
+DenseMap<const Type *, Function *> m_ndfn;
+
+// TODO: extract into common library for nondet pass and this code
+Function &createNewNondetFn(Module &m, Type &type, unsigned num,
+                            std::string prefix) {
+  std::string name;
+  unsigned c = num;
+
+  do
+    name = prefix + std::to_string(c++);
+  while (m.getNamedValue(name));
+  Function *res =
+      dyn_cast<Function>(m.getOrInsertFunction(name, &type).getCallee());
+  assert(res);
+  return *res;
+}
+
+// TODO: extract into common library for nondet pass and this code
+Function *getNondetFn(Type *type, Module *m) {
+  auto it = m_ndfn.find(type);
+  if (it != m_ndfn.end()) {
+    return it->second;
+  }
+
+  Function *res =
+      &createNewNondetFn(*m, *type, m_ndfn.size(), "verifier.nondet.");
+  m_ndfn[type] = res;
+  return res;
+}
+
+// Replace the given function body with code that stores ND values
+// in output args.
+void replaceFnBodyWithND(Function *oldfn, SetVector<Value *> &inputs,
+                         SetVector<Value *> &outputs) {
+  Function *TheFunction = oldfn;
+  auto ret_ty = TheFunction->getReturnType();
+  TheFunction->dropAllReferences(); // delete body of function
```

Comment:
Everything looks ok, except I'm worried about this `dropAllReferences`. If you look at `Function::dropAllReferences` implementation, the first few lines delete all basic blocks, then other stuff is deleted as well. I don't know whether we want the other stuff deleted or kept.

However, given that this is experimental code, I'm ok with leaving it as is.

Reply:
I see. I used `dropAllReferences` because it is used by `Function::deleteBody` which drops all refs and sets linkage to external. We don't want the second step.

https://sourcegraph.com/github.com/llvm/llvm-project@release/14.x/-/blob/llvm/include/llvm/IR/Function.h?L676#tab=def